pkg/pool-manager/pool.go

```diff
@@ -186,31 +195,97 @@ func getNextMac(currentMac net.HardwareAddr) net.HardwareAddr {
 	return currentMac
 }
 
-// Checks if the namespace of an instance is opted in for kubemacpool
-func (p *PoolManager) isInstanceOptedIn(namespaceName, mutatingWebhookConfigName, webhookName string) (bool, error) {
+// compareNamespaceSelectorToOptMode decides whether a namespace should be managed
+// by comparing the mutating-webhook's namespaceSelector (that defines the opt-mode)
+// and its compatibility to the given label in the namespace
+func (p *PoolManager) compareNamespaceSelectorToOptMode(namespaceName, mutatingWebhookConfigName, webhookName string, vmOptMode OptMode) (bool, error) {
+	isNamespaceManaged, err := isNamespaceManagedByDefault(vmOptMode)
+	if err != nil {
+		return false, errors.Wrap(err, "Failed to check if namespaces are managed by default by opt-mode")
+	}
+
 	ns, err := p.kubeClient.CoreV1().Namespaces().Get(context.TODO(), namespaceName, metav1.GetOptions{})
 	if err != nil {
-		return false, errors.Wrapf(err, "Failed to get Namespace %s", namespaceName)
+		return false, errors.Wrap(err, "Failed to get Namespace")
 	}
 	namespaceLabelMap := ns.GetLabels()
-	log.V(3).Info("namespaceName Labels", "namespaceName", namespaceName, "Labels", namespaceLabelMap)
+	log.V(1).Info("namespaceName Labels", "vm instance namespaceName", namespaceName, "Labels", namespaceLabelMap)
 
 	mutatingWebhookConfiguration, err := p.kubeClient.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(context.TODO(), mutatingWebhookConfigName, metav1.GetOptions{})
 	if err != nil {
-		return false, errors.Wrapf(err, "Failed to get mutatingWebhookConfig %s", mutatingWebhookConfigName)
+		return false, errors.Wrap(err, "Failed to get mutatingWebhookConfig")
 	}
 
 	for _, webhook := range mutatingWebhookConfiguration.Webhooks {
 		if webhook.Name == webhookName {
```

Comment:
Based on the previous answer to my question, I understand this is expected to occur only once, therefore once it is matched there is no point to go over the rest of the webhooks. (I think this is what originally happens)

It will also be nice to express that this name is expected only once by splitting the lookup from the operation on it:
```
webhook = lookupWebhook(webhookName) // replacing the for-loop
if namespaceLabelSet := labels.Set(namespaceLabelMap); namespaceLabelSet != nil {
    return isNamespaceManagedByWebhookNamespaceSelector(webhook, vmOptMode, namespaceLabelSet, isNamespaceManaged)
}
```

Reply:
I agree. Done